integrations/go/workload-executor.go

```diff
@@ -0,0 +1,315 @@
+	if result == nil {
+		return true
+	}
+
+	var expected struct {
+		MatchedCount  int64 `bson:"matchedCount"`
+		ModifiedCount int64 `bson:"modifiedCount"`
+		UpsertedCount int64 `bson:"upsertedCount"`
+	}
+	err := bson.Unmarshal(result.(bson.Raw), &expected)
+	if err != nil {
+		return false
+	}
+
+	if expected.MatchedCount != res.MatchedCount {
+		return false
+	}
+	if expected.ModifiedCount != res.ModifiedCount {
+		return false
+	}
+
+	actualUpsertedCount := int64(0)
+	if res.UpsertedID != nil {
+		actualUpsertedCount = 1
+	}
+	return expected.UpsertedCount == actualUpsertedCount
+}
+
+func executeCollectionOperation(coll *mongo.Collection, op *operation) (bool, error) {
+	switch op.Name {
+	case "insertOne":
+		res, err := executeInsertOne(coll, op.Arguments)
+		return verifyInsertOneResult(res, op.Result), err
+	case "find":
+		cursor, err := executeFind(coll, op.Arguments)
+		return verifyCursorResult(cursor, op.Result), err
+	case "updateOne":
+		res, err := executeUpdateOne(coll, op.Arguments)
+		return verifyUpdateResult(res, op.Result), err
+	}
+	return false, errors.New("unrecognized collection operation: " + op.Name)
+}
+
+func runOperation(coll *mongo.Collection, op *operation) (bool, error) {
+	// execute the command on the given object
+	if op.Object == "collection" {
+		return executeCollectionOperation(coll, op)
+	}
+	return false, errors.New("unrecognized object: " + op.Name)
+}
+
+func main() {
+	connstring := os.Args[1]
+	workloadSpec := os.Args[2]
+
+	var workload driverWorkload
+
+	err := bson.UnmarshalExtJSONWithRegistry(specTestRegistry,[]byte(workloadSpec), false, &workload)
+	if err != nil {
+		str := "failed to unmarshal data: " + err.Error()
+		panic(str)
+	}
+
+	client, err := mongo.Connect(context.Background(), options.Client().ApplyURI(connstring))
+	if err != nil {
+		str := "failed to connect to client: " + err.Error()
+		panic(str)
+	}
+	defer func() { _ = client.Disconnect(context.Background()) }()
+
+	db := client.Database(workload.Database)
+	coll := db.Collection(workload.Collection)
+
+	results := struct {
+		NumErrors    int `json:"numErrors"`
+		NumFailures  int `json:"numFailures"`
+		NumSuccesses int `json:"numSuccesses"`
+	}{}
+
+	done := make(chan struct{})
+
+	go func() {
+		c := make(chan os.Signal, 1)
+		signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+
+		<-c
+		close(done)
+	}()
+
+	_ = coll.Drop(context.Background())
+
+	// insert testdata
+	if len(workload.TestData) > 0 {
+		_, err := coll.InsertMany(context.Background(), workload.TestData)
+		if err != nil {
+			str := "inserting testData failed: " + err.Error()
+			panic(str)
+		}
+	}
+
+	defer func() {
+		data, err := json.Marshal(results)
+		if err != nil {
+			str := "marshal results failed: " + err.Error()
+			panic(str)
+		}
+		path, _ := os.Getwd()
+		err = ioutil.WriteFile(path+"/results.json", data, 0644)
+		if err != nil {
+			str := "write to file failed: " + err.Error()
+			panic(str)
+		}
+	}()
+
+	for {
```

Comment:
Why do we need a nested loop here? Is this equivalent to:

```
// do opportunistic check for done
select {
     case <-done:
         return
     default:
}

// Run ops, checking for done before each op
for _, operation : = range operations {
    // same select as above with done/default cases
    // runOp
}

// Wait for done before returning
// select with single done case, no default

Reply:
i cleaned it up, but we do want it to repeat the operations from the top until astrolabe ends it